p2p/lightchain.py

```diff
@@ -136,22 +149,90 @@ def callback(r: protocol._DecodedMsgType) -> None:
+            rlp_account = HexaryTrie.get_from_proof(header.state_root, key, proof)
+        except BadTrieProof as exc:
+            raise BadLESResponse("Peer %s returned an invalid proof for account %s at block %s" % (
+                peer,
+                encode_hex(address),
+                encode_hex(block_hash),
+            )) from exc
         return rlp.decode(rlp_account, sedes=Account)
 
     @alru_cache(maxsize=1024, cache_exceptions=False)
-    async def get_contract_code(self, block_hash: Hash32, key: bytes) -> bytes:
-        peer = cast(LESPeer, self.peer_pool.highest_td_peer)
+    @service_timeout(COMPLETION_TIMEOUT)
+    async def get_contract_code(self, block_hash: Hash32, address: Address) -> bytes:
+        """
+        :param block_hash: find code as of the block with block_hash
+        :param address: which contract to look up
+
+        :return: bytecode of the contract, ``b''`` if no code is set
+
+        :raise NoEligiblePeers: if no peers are available to fulfill the request
+        :raise TimeoutError: if an individual request or the overall process times out
+        """
+        # get account for later verification, and
+        # to confirm that our highest total difficulty peer has the info
+        try:
+            account = await self.get_account(block_hash, address)
+        except HeaderNotFound as exc:
+            raise NoEligiblePeers("Our best peer does not have header %s" % block_hash) from exc
+
+        code_hash = account.code_hash
+
+        return await self._retry_on_bad_response(
+            partial(self._get_contract_code_from_peer, block_hash, address, code_hash)
+        )
+
+    async def _get_contract_code_from_peer(
+            self,
+            block_hash: Hash32,
+            address: Address,
+            code_hash: Hash32,
+            peer: LESPeer) -> bytes:
+        """
+        A single attempt to get the contract code from the given peer
+
+        :raise BadLESResponse: if the peer replies with contract code that does not match the
+            account's code hash
+        """
+        # request contract code
         request_id = gen_request_id()
-        peer.sub_proto.send_get_contract_code(block_hash, key, request_id)
+        peer.sub_proto.send_get_contract_code(block_hash, keccak(address), request_id)
         reply = await self._wait_for_reply(request_id)
+
         if not reply['codes']:
-            return b''
-        return reply['codes'][0]
+            bytecode = b''
+        else:
+            bytecode = reply['codes'][0]
+
+        # validate bytecode against a proven account
+        if code_hash == keccak(bytecode):
+            return bytecode
+        elif bytecode == b'':
+            # TODO disambiguate failure types here, and raise the appropriate exception
+            # An (incorrectly) empty bytecode might indicate a bad-acting peer, or it might not
+            raise NoEligiblePeers("Our best peer incorrectly responded with an empty code value")
```

Comment:
I'm not sure if I'm splitting hairs but this exception seems wrong.  If the code came into existence at block 5, and our best peer is at block 10, but some of our other peers are at blocks 6 and 8, then we still have eligible peers, but there is something wrong with our *best* peers.

Mostly, my issue is with the name of the exception and whether we need some other exception to capture the concept of what has gone wrong here.